### PR TITLE
Support arbitrary `CtrlSpec` to single qubit control

### PR DESCRIPTION
**Pull request by anurudhp**
towards #1131 

- Introduces `CtrlSpecAnd` which is a generalized And/MultiAnd for an arbitrary `CtrlSpec`.

To follow:
- Fallback for `Controlled.decompose_bloq` to use this, followed by single-controlled subbloq, and uncompute.

### CODE REVIEW

File: qualtran/bloqs/mcmt/ctrl_spec_activation_test.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    _ctrl_on_nd_bits,
+    CtrlSpecActivation,
+)
+
+
+@pytest.mark.parametrize(
+    "example", [_ctrl_on_bits, _ctrl_on_nd_bits, _ctrl_on_int, _ctrl_on_multiple_values]
+)
+def test_examples(bloq_autotester, example):
+    bloq_autotester(example)
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_activation_test.py:30]: Please add correctness tests via classical / tensor simulation.

File: qualtran/bloqs/mcmt/ctrl_spec_activation.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+
+@frozen
+class CtrlSpecActivation(Bloq):
+    """Computes a single qubit which is `1` iff the CtrlSpec is satisfied.
+
+    This reduces an arbitrary control specification to a single qubit, which can be used
+    to then control a bloq. Therefore, a bloq author is only required to implement a
+    single-controlled version of their bloq, and can be generalized to arbitrary controls.
+
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_activation.py:49]: This has a caveat that the current implementation works only for the case when the `CtrlSpec` is a logical `AND` of all individual control register clauses and each control register clause is restricted to be an `AND` with a single constant. 

Users can derive `CtrlSpec` to provide more general control spec and `activation_function_dtypes` and `is_active` methods were designed to support this. Although the docstring of `CtrlSpec` explicitly states

```
CtrlSpec uses logical AND among all control register clauses. If you need a different boolean
    function, open a GitHub issue.
```

For this PR, we should

a) Document this restriction in the docstring. 
b) Maybe rename to `CtrlSpecAND` (or something similar) which is more indicative of this restriction and we explicitly state that we don't support constructing a circuit for arbitrary boolean functions that a user may encode as part of the `CtrlSpec`.

**Review by tanujkhattar — COMMENTED** [review]
Since we are decomposing a `CtrlSpec` to an `AND` gate, I'd suggest renaming to `CtrlSpecAND`. See my detailed comment above. 

Looking good otherwise.

File: qualtran/bloqs/mcmt/ctrl_spec_activation_test.py (hunk 1 above)
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_activation_test.py:32]: added classical simulation test.

**Comment by mpharrigan** [comment]
If we go with `CtrlSpecAnd` I suggest following the capitalization of existing `And`

File: qualtran/bloqs/mcmt/ctrl_spec_activation.py (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_activation.py]: Updated the bloq name and the docstring explaining this restriction

File: qualtran/bloqs/bookkeeping/partition.py [hunk 3]
```diff
@@ -137,22 +138,24 @@ def _classical_partition(self, x: 'ClassicalValT') -> Dict[str, 'ClassicalValT']
...
 
-    def _classical_unpartition(self, **vals: 'ClassicalValT'):
-        out_vals = []
+    def _classical_unpartition_to_bits(self, **vals: 'ClassicalValT') -> NDArray[np.uint]:
+        out_vals: list[list[int]] = []
         for reg in self.regs:
             reg_val = vals[reg.name]
             if isinstance(reg_val, np.ndarray):
-                out_vals.append(ints_to_bits(reg_val.ravel(), reg.bitsize).ravel())
+                out_vals.extend(reg.dtype.to_bits(val) for val in reg_val.ravel())
```
Comment (tanujkhattar) [inline qualtran/bloqs/bookkeeping/partition.py:146]: Why this change? Also, why this change only in the `_classical_unpartition` and not in `_classical_partition` ? Was the previous logic wrong for certain cases? Can you please add tests which pass with the new version and fail with the original one?

**Review by tanujkhattar — APPROVED** [review]
LGTM % comments about partition

File: qualtran/bloqs/bookkeeping/partition.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/bookkeeping/partition.py:146]: I'm using this to convert the `cvs` to bits in `CtrlSpecAnd` which allows any dtype. But the current classical simulation protocol only supports integer values in `ClassicalValT`, so seems like it worked either way.

Comment (anurudhp) [reply qualtran/bloqs/bookkeeping/partition.py:146]: Also `bits_to_ints` / `ints_to_bits` are faster than manually calling `to_bits` / `from_bits` on each element. (similar to https://github.com/quantumlib/Qualtran/issues/1072). So only changed the part I'm using for now.

Comment (tanujkhattar) [reply qualtran/bloqs/bookkeeping/partition.py:146]: Sorry, I don't understand. What was the part which was failing earlier and now gets fixed? I presume even now, `CtrlSpecAnd` would support only `QUInt`, `QInt` and `QAny` types (or, types for which `bits_to_ints` works correctly) since the classical bits need to be partitioned correctly (right now, partitioning is still based on `bits_to_ints` but unpartitioning uses `dtype.to_bits()`) ?

Comment (anurudhp) [reply qualtran/bloqs/bookkeeping/partition.py:146]: `CtrlSpecAnd` also works for `QFxp` as it's just bitwise equality

Comment (anurudhp) [reply qualtran/bloqs/bookkeeping/partition.py:146]: I'm using this only for extracting the classical control bits in `CtrlSpecAnd` (i.e. `_flat_cvs`), so I never merge these bits back to get the constants. 

But this would break the classical simulator, but as the classical sim only supports ints now, we don't have any tests that use Fxp.

File: qualtran/bloqs/mcmt/ctrl_spec_and_test.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+@pytest.mark.parametrize("ctrl_spec", [CtrlSpec(), CtrlSpec(cvs=0), CtrlSpec(cvs=[0])])
+def test_raises_for_single_qubit_controls(ctrl_spec: CtrlSpec):
+    with pytest.raises(ValueError):
+        _ = CtrlSpecAnd(ctrl_spec)
+
+
+@pytest.mark.parametrize(
+    "ctrl_spec",
+    [CtrlSpec(QUInt(4), cvs=4), CtrlSpec((QUInt(4), QUInt(4)), cvs=(np.array(3), np.array(5)))],
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_and_test.py:43]: will add a tensor test for it, as `get_classical_truth_table` uses classical sim which doesn't support Fxp yet.

File: qualtran/bloqs/bookkeeping/partition.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/bloqs/bookkeeping/partition.py:146]: I think for now, it's better to say that `CtrlSpecAnd` does not support `QFxp`, because in reality it does not. The decomposition would be incorrect because `Partition` and `Unpartition` doesn't support `QFxp`. So it's fine to fail early at `_flat_cvs` instead of having a wrong decomposition. 

The alternative would be to compute the `_flat_cvs` without invoking `Partition` so that the computation of `_flat_cvs` works for all data types and update `Partition` bloq for `_classical_unpartition` and `_classical_partition` whenever we are ready to do so. 

The current logic leaves the state of things in a complicatedly broken way where one direction works (maybe) but the other doesn't.

Comment (anurudhp) [reply qualtran/bloqs/bookkeeping/partition.py]: I reverted these changes. For now CtrlSpecAnd only supports integer types (checked in post_init). 

Will add more support after the classical sim for QFxp is merged.

File: qualtran/bloqs/mcmt/ctrl_spec_and_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_and_test.py:43]: not supported in this PR

**Review by tanujkhattar — APPROVED** [review]
